enterprise/server/remote_execution/uffd/uffd.go

```diff
@@ -399,25 +399,46 @@ func (h *Handler) handlePageFault(uffd uintptr, memoryStore *copy_on_write.COWSt
 	}
 
 	// If address had been previously removed, zero it.
-	// TODO: If there are sequential removed pages, zero multiple pages at once.
 	if _, removed := h.removedAddresses[int64(guestPageAddr)]; removed {
-		delete(h.removedAddresses, int64(guestPageAddr))
+		return h.copyZeroes(uffd, guestPageAddr)
+	}
+
+	// Otherwise copy data from the memory snapshot.
+	return h.copyDataFromMemorySnapshot(uffd, memoryStore, mapping, guestPageAddr)
+}
+
+// copyZeroes handles a page fault by copying zeroes into the guest. The guest
+// expects zeroes if the faulting page was previously removed by the balloon.
+func (h *Handler) copyZeroes(uffd uintptr, faultingAddress uintptr) error {
+	// If multiple consecutive pages were removed, zero them all.
+	end := faultingAddress
+	for {
+		_, removed := h.removedAddresses[int64(end)]
+		if !removed {
+			break
+		}
 
+		end += uintptr(os.Getpagesize())
```

Comment:
I sort of forget how the `guestMemoryAddrToMapping` stuff works - do we need to worry about going past the end of the `mapping`?

Reply:
We use the mappings when we need to copy data from a snapshot file to the guest. It translates snapshot file offsets to guest memory addresses. Because we're just zeroing pages in the guest here, we don't need to use the mappings